File: C#/GenericStructures/List.cs
using System;

class List<T>{
    class Node{
        public T data;
        public Node prev = null 
        public Node next = null;
        public Node(T data){
            this.data = data;
        }
    }

    private Node front = null;
    private Node back = null;
    private int m_size;

    public List(){
        m_size = 0;
    }

    public void add(T data){
        Node node = new Node(data);
        if(is_empty()){
            this.front = node;
            this.back = node;
        }
        else{
            this.back.next = node;
            node.prev = this.back;
            this.back = node;
        }
        ++ this.m_size;
    }

    public void add_front(T data){
        Node node = new Node(data);
        
        if(is_empty()){
            this.front = node;
            this.back = node;
        }else{
            node.next = this.front;
            this.front.prev = node;
            this.front = node;
        }
        this.m_size += 1;
    }

    public void add_back(T data){
        Node node = new Node(data);
        
        if(is_empty()){
            this.front = node;
            this.back = node;
        }else{
            node.prev = this.back;
            this.back.next = node;
            this.back = node;            
        }
        this.m_size += 1;
    }

    public void remove(T data){
        if(is_empty()){
            // error
        }else{
            Node ptr = this.front;

            while(ptr.next != null){
                if(ptr.data == data){
                    ptr.prev = ptr.prev.prev;
                    ptr.next = ptr.next.next;

                    this.m_size -= 1;
                    return;
                }
                ptr = ptr.next;
            }
        }
    }

    private void remove_front(){
        try{
            if(is_empty()) throw System.Exception;
            Node temp = this.front; 
            this.front = front.next;
            temp = null;

            this.m_size -= 1;
        }
        catch (System.Exception){
            throw; // handle exception
        }
    }

    private void remove_back(){
        try{
            if(is_empty()) throw System.Exception;
            Node temp = this.back; 
            this.back = front.prev;
            temp = null;

            this.m_size -= 1;
        }
        catch (System.Exception){
            throw; // handle exception
        }
    }

    public void display(){
        Node ptr = this.front;
        while(ptr.next != null){
            Console.Write(ptr.data + " ");
            ptr = ptr.next;
        }
    }
    
    public void display_reverse(){
        Node ptr = this.back;
        while(ptr.prev != null){
            ptr = ptr.prev;
            Console.Write(ptr.data + " ");
        }
    }

    public bool is_empty(){
        return this.front == null && this.back == 0;
    }
}

class MainDoubly{
    public static void Main(string[] args){
        List dl = new List();
        for(int i = 0; i<10; ++i)
            dl.add(i);
        
        dl.display();
        Console.WriteLine("\nprints backwards");
        dl.display_reverse();
        
    }
}

File: Python/GenericStructures/List.py
from typing import TypeVar, Generic

T = TypeVar("T")


class Node(Generic[T]):
    next = None
    prev = None

    def __init__(self, data: T = None) -> None:
        self.data: T = data


class List(Generic[T]):
    front: Node[T] = None
    back: Node[T] = None
    size: int = 0

    def add_back(self, data: int) -> None:
        node = Node[T](data)
        if self.is_empty():
            self.front = node
            self.back = node
        else:
            self.back.next = node
            node.prev = self.back
            self.back = node

        self.size += 1

    def add_front(self, data: int) -> Node:
        node = Node[T](data)
        if self.is_empty():
            self.front = node
            self.back = node
        else:
            self.front.prev = node
            node.next = self.node
            self.front = node

        self.size += 1

    def remove(self, data: int) -> None:
        if self.is_empty():
            raise Exception("Error: element not found in the list.")

        elif self.front.data == data:
            self._remove_front()
        elif self.back.data == data:
            self._remove_back()
        else:
            ptr: Node[T] = self.front
            prev: Node[T] = ptr

            while ptr.next is not None:
                if ptr.data == data:
                    prev.next = ptr.next
                    ptr.next.prev = prev

                    ptr = None
                    self.size -= 1
                    return
                prev = ptr
                ptr = ptr.next

    def display(self) -> None:
        ptr: Node[T] = self.front
        while ptr.next is not None:
            print(ptr.data)
            ptr = ptr.next

    def _remove_front(self) -> None:
        temp: Node[T] = self.front
        self.front = self.front.next
        temp = None
        self.size -= 1

    def _remove_back(self) -> None:
        temp: Node[T] = self.back
        self.back = self.back.prev
        temp = None
        self.size -= 1

    def is_empty(self) -> bool:
        return self.front is None and self.back is None

    def display_reverse(self) -> None:
        ptr: Node[T] = self.back
        while ptr.next is not None:
            print(ptr.data)
            ptr = ptr.next


if __name__ == "__main__":
    l = List[int]()
    for i in range(1, 10):
        l.add_back(i)

    l.remove(3)
    l.display_reverse()
